core/commands/diff.py

```diff
@@ -174,51 +178,165 @@ def run(self, edit, cursors=None, navigate_to_next_hunk=False):
+            "Showing {} changes".format("staged" if next_mode else "unstaged")
+        )
 
         self.view.run_command("gs_diff_refresh")
+
+        just_hunked = self.view.settings().get("git_savvy.diff_view.just_hunked")
+        # Check for `last_cursors` as well bc it is only falsy on the *first*
+        # switch. T.i. if the user hunked and then switches to see what will be
+        # actually comitted, the view starts at the top. Later, the view will
+        # show the last added hunk.
+        if just_hunked and last_cursors:
+            self.view.settings().set("git_savvy.diff_view.just_hunked", "")
+            region = find_hunk_in_view(self.view, just_hunked)
+            if region:
+                set_and_show_cursor(self.view, region.a)
+                return
+
         if last_cursors:
-            sel = self.view.sel()
-            sel.clear()
-            for (a, b) in last_cursors:
-                sel.add(sublime.Region(a, b))
-            self.view.show(sel)
+            # The 'flipping' between the two states should be as fast as possible and
+            # without visual clutter.
+            with no_animations():
+                set_and_show_cursor(self.view, unpickle_sel(last_cursors))
+
+
+def find_hunk_in_view(view, hunk):
+    hunk_content = extract_first_hunk(hunk)
+    if hunk_content:
+        return (
+            view.find(hunk_content[0], 0, sublime.LITERAL)
+            or search_for_hunk_content_in_view(view, hunk_content[1:])
+        )
+
+
+def extract_first_hunk(hunk):
+    hunk_lines = hunk.split('\n')
+    not_hunk_start = lambda l: not l.startswith('@@ ')
+
+    try:
+        start, *rest = dropwhile(not_hunk_start, hunk_lines)
+    except (StopIteration, ValueError):
+        return None
+
+    return [start] + list(takewhile(not_hunk_start, rest))
+
+
+def search_for_hunk_content_in_view(view, lines):
+    for hunk_content in shrink_list_sym(lines):
+        region = view.find('\n'.join(hunk_content), 0, sublime.LITERAL)
+        if region:
+            return first_hunk_start_before_pt(view, region.a)
+
+
+def first_hunk_start_before_pt(view, pt):
+    for region in line_regions_before_pt(view, pt):
+        if view.substr(region).startswith('@@ '):
+            return region
+
+
+def shrink_list_sym(list):
+    while list:
+        yield list
+        list = list[1:-1]
+
+
+def line_regions_before_pt(view, pt):
+    row, _ = view.rowcol(pt)
+    for row in reversed(range(row)):
+        pt = view.text_point(row, 0)
+        yield view.line(pt)
+
+
+def pickle_sel(sel):
+    return [(s.a, s.b) for s in sel]
+
+
+def unpickle_sel(pickled_sel):
+    return [sublime.Region(a, b) for a, b in pickled_sel]
+
+
+def set_and_show_cursor(view, cursors):
+    sel = view.sel()
+    sel.clear()
+    try:
+        it = iter(cursors)
+    except TypeError:
+        sel.add(cursors)
+    else:
+        for c in it:
+            sel.add(c)
+
+    view.show(sel)
+
+
+@contextmanager
+def no_animations():
```

Comment:
In a later version I will get rid of this side-effect by just storing and restoring the `viewport_position`.